pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -394,35 +434,14 @@ func (configgen *ConfigGeneratorImpl) buildInboundClusters(cb *ClusterBuilder, p
 		if noneMode {
 			return nil
 		}
-
-		clustersToBuild := make(map[int][]*model.ServiceInstance)
-		for _, instance := range instances {
-			// For service instances with the same port,
-			// we still need to capture all the instances on this port, as its required to populate telemetry metadata
-			// The first instance will be used as the "primary" instance; this means if we have an conflicts between
-			// Services the first one wins
-			ep := int(instance.Endpoint.EndpointPort)
-			clustersToBuild[ep] = append(clustersToBuild[ep], instance)
-		}
-
-		bind := actualLocalHost
-		if features.EnableInboundPassthrough {
-			bind = ""
-		}
-		// For each workload port, we will construct a cluster
-		for epPort, instances := range clustersToBuild {
-			// The inbound cluster port equals to endpoint port.
-			localCluster := cb.buildInboundClusterForPortOrUDS(epPort, bind, proxy, instances[0], instances)
-			// If inbound cluster match has service, we should see if it matches with any host name across all instances.
-			hosts := make([]host.Name, 0, len(instances))
-			for _, si := range instances {
-				hosts = append(hosts, si.Service.Hostname)
-			}
-			clusters = cp.conditionallyAppend(clusters, hosts, localCluster.build())
-		}
+		clusters = configgen.buildClustersFromServiceInstances(cb, proxy, instances, cp, false)
```

Comment:
nit: if buildClustersFromServiceInstances is going to allocate its own array, then don't allocate one on L417

Reply:
I did double check. Wont work. Check L514-516; to merge from L443.

Reply:
Other option is to move to L440 as its needed for L442 but since its var would be good to keep it at L417 as is since its existing code only.